django/utils/choices.py

```diff
@@ -0,0 +1,51 @@
+from collections.abc import Callable, Iterable, Iterator, Mapping
+
+from django.db.models.enums import ChoicesMeta
+from django.utils.functional import Promise
+
+
+class ChoiceIterator:
+    """Base class for lazy iterators for choices."""
+
+
+class CallableChoiceIterator(ChoiceIterator):
+    """Iterator to lazily normalize choices generated by a callable."""
+
+    def __init__(self, choices_func):
+        self.choices_func = choices_func
+
+    def __iter__(self):
+        yield from normalize_field_choices(self.choices_func())
+
+
+def normalize_field_choices(value, *, depth=0):
+    """Normalizes choices values consistently for fields and widgets."""
+
+    match value:
+        case ChoicesMeta():
+            # Choices enumeration helpers already produce output in canonical form.
+            return value.choices
+        case ChoiceIterator() | Promise() | bytes() | str():
+            # Avoid expanding lazy choice iterators too soon by returning early.
+            # String-like types are iterable, so return early to avoid iterating.
+            return value
+        case Callable() if depth == 0:
+            # If at the top level, wrap callables to be evaluated lazily.
+            return CallableChoiceIterator(value)
```

Comment:
> I do like the idea of support lazy iterator even on nested choices, but I'd have that being a follow up branch and not within this one.

Yes, perhaps this would be clearer as the following:

```suggestion
        case Callable() if depth == 0:
            # If at the top level, wrap callables to be evaluated lazily.
            return CallableChoiceIterator(value)
        case Callable():  # depth > 0
            value = value()
```

Note that we only want to be lazy at the top level. It's pointless to be lazy for the nested callables as we'd just want to resolve them all at the same time to generate the choices all at once.

> If I comment out the if depth == 0 from the match case, no test fails. Do you think you could add a test for that?

Yeah, so I'm going to add a fixup to handle the nested callable case. The reason is that currently, with the above test we have the following failure:

```
======================================================================
FAIL [0.001s]: test_nested_callable (utils_tests.test_choices.NormalizeFieldChoicesTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/utils_tests/test_choices.py", line 110, in test_nested_callable
    self.assertEqual(list(output), self.expected_nested)
AssertionError: Lists differ: [('Audio', functools.partial(<function NormalizeFiel[257 chars]wn')] != [('Audio', [('vinyl', 'Vinyl'), ('cd', 'CD')]), ('Vi[64 chars]wn')]

First differing element 0:
('Audio', functools.partial(<function NormalizeFiel[87 chars]io'))
('Audio', [('vinyl', 'Vinyl'), ('cd', 'CD')])

+ [('Audio', [('vinyl', 'Vinyl'), ('cd', 'CD')]),
+  ('Video', [('vhs', 'VHS Tape'), ('dvd', 'DVD')]),
- [('Audio',
-   functools.partial(<function NormalizeFieldChoicesTests.test_nested_callable.<locals>.get_choices at 0x7f64f8d14940>, medium='audio')),
-  ('Video',
-   functools.partial(<function NormalizeFieldChoicesTests.test_nested_callable.<locals>.get_choices at 0x7f64f8d14940>, medium='video')),
   ('unknown', 'Unknown')]
```

Which is basically broken anyway because the nested case was never properly handled.

And removing the `if depth == 0` fails in this way:

```
======================================================================
FAIL [0.001s]: test_nested_callable (utils_tests.test_choices.NormalizeFieldChoicesTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/utils_tests/test_choices.py", line 110, in test_nested_callable
    self.assertEqual(list(output), self.expected_nested)
AssertionError: Lists differ: [('Audio', <django.utils.choices.CallableChoiceItera[133 chars]wn')] != [('Audio', [('vinyl', 'Vinyl'), ('cd', 'CD')]), ('Vi[64 chars]wn')]

First differing element 0:
('Audio', <django.utils.choices.CallableChoiceItera[25 chars]730>)
('Audio', [('vinyl', 'Vinyl'), ('cd', 'CD')])

+ [('Audio', [('vinyl', 'Vinyl'), ('cd', 'CD')]),
+  ('Video', [('vhs', 'VHS Tape'), ('dvd', 'DVD')]),
- [('Audio',
-   <django.utils.choices.CallableChoiceIterator object at 0x7ff75319f730>),
-  ('Video',
-   <django.utils.choices.CallableChoiceIterator object at 0x7ff75319f820>),
   ('unknown', 'Unknown')]
```

Note the `CallableChoiceIterator` there instead. This is definitely wrong and why the guard was put in place as nested callables shouldn't be lazy - that makes no sense.

But, yes, without resolving the nested callables, this doesn't make much sense. Fix incoming...

Reply:
Excellent, thank you!